### PR TITLE
Codify Coding Standards

### PR DESCRIPTION
**Pull request by jcupler**
### Description
Codifies team coding standards, to hopefully establish shared standards and explicitly define assumptions.

### CODE REVIEW

File: docs/coding_standards.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        - 
+    - 
+    - 
+- All test code and packages SHOULD live in `/src/test/**`
+    - The structure of unit testing code SHOULD mirror the application code exactly, starting from `/src/test/unit/**`
+    - The structure of functional testing SHOULD be packaged based on the what function they target, starting from `/src/test/functional/**`
+    - The structure of integration testing SHOULD be packaged based on their target, and SHOULD be ran against documentation standards and intended target, starting from `/src/test/integration/**`
+
+### REST standards
+- REST endpoints SHALL be versioned
```
Comment (kyperbelt) [inline docs/coding_standards.md:58]: I like this: can we add an example so people know what we mean by versioning? for example, `v1/endpoint/{resource}`

File: docs/coding_standards.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ - MAY, OPTIONAL are to be used to establish OPTIONAL criteria whose implentation, or lack there of, is left to the developer. This MUST NOT effect core functionality.
+ - The above interpretations SHALL only hold when the given word is fully capitalized
+
+## Git and Version control standards
+- Developers MUST NOT push directly to main. Instead, developers MUST develop feature branches, which are to be merged into main after review.
+    - There SHALL be one exception: Developers can force push to main, for extreme circumstances. Usage of this exception SHOULD be avoided, and developers SHALL notify the team upon usage
+- Developers SHALL be subject to a code review before feature branches are to be merged to main.
+    - Specfically, there MUST be at least one approval before a branch can be merged to main.
+    - There SHALL be one exception: Developers can force push to main without review, for extreme circumstances. Usage of this exception SHOULD be avoided, and developers SHALL notify the team upon usage
+- Developers SHALL format their commit messages using [yaml](https://learnxinyminutes.com/docs/yaml/) syntax.
```
Comment (kyperbelt) [inline docs/coding_standards.md:33]: I don't mind doing this, but can you explain why we need it?

File: docs/coding_standards.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+These standards establish the following:
+ - MUST, REQUIRED, and SHALL are to be used to establish hard requirements that MUST be met to meet acceptence criteria.
+ - MUST NOT, SHALL NOT are to be used to establish hard requirements that MUST be avoided in order to meet acceptence criteria.
+ - SHOULD, RECOMMENDED are to be used to establish soft requirements that SHOULD be meet, but deviations are acceptable provided there is a good reason.
+ - SHOULD NOT, NOT RECOMMENDED to be used to establish soft requirements that SHOULD be avoided, but deviations are acceptable provided there is a good reason.
+ - MAY, OPTIONAL are to be used to establish OPTIONAL criteria whose implentation, or lack there of, is left to the developer. This MUST NOT effect core functionality.
+ - The above interpretations SHALL only hold when the given word is fully capitalized
+
+## Git and Version control standards
+- Developers MUST NOT push directly to main. Instead, developers MUST develop feature branches, which are to be merged into main after review.
```
Comment (kyperbelt) [inline docs/coding_standards.md:28]: Can we add something here for deleting feature branches from the repo once they are merged or closed please?

File: docs/coding_standards.md
Comment (kyperbelt) [inline docs/coding_standards.md:1]: Overall, great job, I support us having this document, Do you think we can add something here to make sure we are moving jira tickets associated with specific commits/prs (if any) to their appropriate place on the Jira board?

**Review by kyperbelt — CHANGES_REQUESTED** [review]

File: docs/coding_standards.md (hunk 1 above)
Comment (jcupler) [reply docs/coding_standards.md:58]: Will do

File: docs/coding_standards.md (hunk 2 above)
Comment (jcupler) [reply docs/coding_standards.md:33]: The commit formatting. I assumed that the parser you've talking about used yaml syntax, so I put it in the standards. WIll change if it uses a different standards.

File: docs/coding_standards.md (hunk 3 above)
Comment (jcupler) [reply docs/coding_standards.md:28]: Could have sworn it's already done. Will do

File: docs/coding_standards.md
Comment (jcupler) [reply docs/coding_standards.md:1]: Will do

**Review by kyperbelt — APPROVED** [review]

**Pull request merged**